ding/model/common/head.py

```diff
@@ -452,6 +452,153 @@ def forward(self, x: torch.Tensor, num_quantiles: Optional[int] = None) -> Dict:
+            >>> assert isinstance(qn_output, torch.Tensor)
+            >>> # default quantile_embedding_size: int = 128,
+            >>> assert qn_output.shape == torch.Size([128, 64])
+        """
+        batch_size = quantiles.shape[0]
+        num_quantiles = quantiles.shape[1]
+        quantile_net = torch.cos(
+            torch.arange(1, self.quantile_embedding_size + 1, 1).view(1, 1, self.quantile_embedding_size).to(quantiles) * math.pi * quantiles.view(batch_size, num_quantiles, 1)
+        ).view(batch_size * num_quantiles, self.quantile_embedding_size)
+        quantile_net = self.iqn_fc(quantile_net)
+        quantile_net = F.relu(quantile_net).view(batch_size, num_quantiles, -1)   #(batch,num_quantiles,hidden_size)
+        return quantile_net
+
+    def forward(self, x: torch.Tensor, num_quantiles: Optional[int] = None) -> Dict:
+        r"""
+        Overview:
+            Use encoded embedding tensor to predict Quantile output.
+            Parameter updates with QuantileHead's MLPs forward setup.
+        Arguments:
+            - x (:obj:`torch.Tensor`):
+                The encoded embedding tensor, determined with given ``hidden_size``, i.e. ``(B, N=hidden_size)``.
+        Returns:
+            - outputs (:obj:`Dict`):
+                Run ``MLP`` with ``QuantileHead`` setups and return the result prediction dictionary.
+
+                Necessary Keys:
+                    - logit (:obj:`torch.Tensor`): Logit tensor with same size as input ``x``.
+                    - q (:obj:`torch.Tensor`): Q valye tensor tensor of size ``(num_quantiles, B, N)``
+                    - quantiles (:obj:`torch.Tensor`): quantiles tensor of size ``(B*num_quantiles, 1)``
+        Examples:
+            >>> head = QuantileHead(64, 64)
+            >>> inputs = torch.randn(4, 64)
+            >>> outputs = head(inputs)
+            >>> assert isinstance(outputs, dict)
+            >>> assert outputs['logit'].shape == torch.Size([4, 64])
+            >>> # default num_quantiles is 32
+            >>> assert outputs['q'].shape == torch.Size([32, 4, 64])
+            >>> assert outputs['quantiles'].shape == torch.Size([128, 1])
+        """
+
+        if num_quantiles is None:
+            num_quantiles = self.num_quantiles
+        batch_size = x.shape[0]
+
+        q_quantiles = self.quantiles_proposal(x.detach())  #(batch,num_quantiles)
+        # accumalative softmax
+        """ accumalative = 1
+        for i in range(1,num_quantiles):
+            q_quantiles[:,num_quantiles-1-i] = accumalative - q_quantiles[:,num_quantiles-i]
+            accumalative = q_quantiles[:,num_quantiles-1-i]
+        q_quantiles[:,-1] = 1 """
+        q_quantiles = torch.cumsum(q_quantiles, dim=1)
+        # quantile_hats: find the optimal condition for τ to minimize W1(Z, τ)  
+        tau_0 = torch.zeros((batch_size, 1)).to(x)
+        q_quantiles_hats = (q_quantiles + torch.cat((tau_0, q_quantiles[:,:-1]), dim=1)).detach() / 2.   #(batch, num_quantiles)
```

Comment:
don't use cat twice like in line 584 and 586, cat only once and use index slice can be better